[WIP] React 16 upgrade

tingard opened the pull request:
Creating a PR to track milestones

Staging branch URL:

Makes the final(?) changes to React 16

List of conflicting modules: 
- [ ] [react-component-resizable](https://www.npmjs.com/package/react-component-resizable)
- [ ] [modal-form](https://www.npmjs.com/package/modal-form)
- [ ] maybe [react-interpolate-component](https://www.npmjs.com/package/react-interpolate-component)
- [ ] ... 

# Required Manual Testing

- [ ] Does the non-logged in home page render correctly?
- [ ] Does the logged in home page render correctly?
- [ ] Does the projects page render correctly?
- [ ] Can you load project home pages?
- [ ] Can you load the classification page?
- [ ] Can you submit a classification?
- [ ] Does talk load correctly?
- [ ] Can you post a talk comment?

# Review Checklist

- [ ] Does it work in all major browsers: Firefox, Chrome, Edge, Safari?
- [ ] Does it work on mobile?
- [ ] Can you `rm -rf node_modules/ && npm install` and app works as expected?
- [ ] If the component is in coffeescript, is it converted to ES6? Is it free of eslint errors? Is the conversion its own commit?
- [ ] Are the tests passing locally and on Travis?

# Optional

- [ ] Have you replaced any `ChangeListener` or `PromiseRenderer` components with code that updates component state?
- [ ] If changes are made to the classifier, does the dev classifier still work?
- [ ] Have you [resized and compressed](https://developers.google.com/web/fundamentals/performance/optimizing-content-efficiency/image-optimization) any image you've added?
- [ ] Have you added in [flow type annotations](https://flowtype.org/docs/type-annotations.html)?
- [ ] Have you followed the [Springer guidelines for commit messages](https://github.com/springernature/frontend-playbook/blob/master/git/git.md#commit-messages)?

srallen commented:
Here's an open PR to update modal-form to use `prop-types` and `create-react-class`: https://github.com/zooniverse/modal-form/pull/33

Is there any chance we don't have to update react-router for React 16? React-router v4's API is pretty awful in my opinion and it'd take a bit of time to update all of the routes to use their new API. Version 3 might work if we do have to upgrade? If not, I have been looking into other routing libraries in a starter test react app.

tingard commented:
Looks like 3.2.0 of react-router will work with react 16 https://github.com/ReactTraining/react-router/issues/5561

tingard commented:
Current failing module is [drag-reorderable](https://www.npmjs.com/package/drag-reorderable)

tingard commented:
So if the idea of JSON subjects goes through I'll probably need to use gzip to keep file sizes down (test file is up to 2.8MB before gzipping and ~350kB after)

eatyourgreens commented:
zooniverse/drag-reorderable#4 removes `React.PropTypes` and upgrades that package to React 15.6.

tingard commented:
Good news is that we're now running tests. Bad news is it's still broken (Markdownz is saying `Cannot read property 'querySelectorAll' of null`)

tingard commented:
from [here](https://reactjs.org/docs/react-dom.html):
> `findDOMNode` only works on mounted components (that is, components that have been placed in the DOM). If you try to call this on a component that has not been mounted yet (like calling `findDOMNode()` in `render()` on a component that has yet to be created) an exception will be thrown.
Maybe should be replaced with refs?

eatyourgreens commented:
Markdownz and the SevenTen client might still be on React 15 here.

eatyourgreens commented:
I'm also seeing this warning when I run the tests, so the `DropdownTask` might be doing something that doesn't work in React 16.

```shell
Warning: Functions are not valid as a React child. This may happen if you return a Component instead of <Component /> from render. Or maybe you meant to call this function rather than return it.
    in div (created by DropdownTask) 
    in div (created by DropdownTask)
    in div (created by GenericTask)
    in GenericTask (created by DropdownTask)
    in DropdownTask (created by WrapperComponent)
    in WrapperComponent
```

eatyourgreens commented:
zooniverse/markdownz#63 should fix the DOM error, and also the same errors that appear in the browser console when you load a new subject.

srallen commented:
@tingard I think react-component-resizable might be cruft from the new project navbar redesign. I might have added it to see how it worked and forgot to uninstall it. I think it can be safely uninstalled.

tingard commented:
Thanks for your comments, I'll have a look into things on Monday!

eatyourgreens commented:
I've tried out zooniverse/markdownz#63 on a local copy of this branch. The tests ran with React warnings but they passed without errors.

eatyourgreens commented:
@tingard the git repo is just the source code. It would still need to be built after downloading.

tingard commented:
@eatyourgreens I'm not getting `src`, so nothing to build. Any ideas why?
All I'm doing is running 
```sh
npm install --save-dev https://github.com/zooniverse/markdownz.git#footnote-links
```
And then checking `node_modules/markdownz`

eatyourgreens commented:
Sorry, I meant you have to clone the repo, build it then install that to PFE. I don't know if it's possible to install it directly from git, since the repo doesn't have the lib directory.

tingard commented:
@eatyourgreens I'm seeing the same (only react warnings)! Thanks for getting that fixed, this is looking close to completion to me?

tingard commented:
Current breaking error: `react-router-scroll` requires `history` ^2.1.2, PFE uses ~4.7.2, errors when starting server.

eatyourgreens commented:
I'd prefer to keep the PropTypes and CreateClass changes separate, since we started those with the upgrade to 15.6, so I've opened #4382.

tingard commented:
Reverted my commit and rebased from master. Any suggestions regarding what to do about react-router-scroll? (rc-slider is fine once bumped to latest as far as I can tell)

tingard commented:
Having removed it from main.cjsx I'm not sure what it was trying to accomplish? Chrome seems to restore scroll itself on some pages

eatyourgreens commented:
It scrolls back to the top on some pages. The exact rules were in `main.cjsx` I think.

tingard commented:
When it's disabled then (afaik) react 16 runs. Though `isMounted` inside promise-render may be problematic.

eatyourgreens commented:
`PromiseRenderer` is an old component that was a mistake in retrospect. There’s some explanation in the Readme.

eatyourgreens commented:
We need that scroll behaviour, by the way, but I don’t remember  which route transitions it was set up for.

tingard commented:
whoops looking at the repo it's just that I haven't bumped to the latest version...

tingard commented:
@eatyourgreens if you have a moment to check this out I can't see anything obviously breaking any more. I should clear up the React warnings on test before merge though.

eatyourgreens commented:
I'll try to make time for this next week.

We should stage it somewhere, and have everyone test it out to see if anything has obviously broken. That's pretty much how we tested the upgrade from React 14 to 15.

tingard commented:
@srallen That was the case on the 22nd of feb. It'll need a rebase and check-through again now though, as new components/tests/package changes might break things. Could be easy but who knows!
I'd do this but am very pressed for time over the next two weeks, and then start a placement.

eatyourgreens commented:
The only merge conflict is the addition of `chart.js` to the package file, so that should be easy to take care of.

srallen commented:
@tingard That's ok. I'm happy to take a look since I'm keen to get upgraded to React 16 too. Could you make sure your fork is editable by us? Here's how to do that: https://help.github.com/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork/

eatyourgreens commented:
I added @tingard to zooniverse on github a few weeks ago. I'll double check that, but you should be able to push directly to this repo. Let me know on Slack if that doesn't work.

tingard commented:
(I'm not going to pretend to fully understand how git forks work)
The branch is pushed to my fork, I have no clue if it's synced to the upstream zooniverse repo but can do that if needed
